### ISSUE TITLE
[BUG] Improper order of export and deprecate declaration for CSVReporter

### ISSUE THREAD

**Issue by yzazik** [google/benchmark#1348]
**Describe the bug**
Redefining `BENCHMARK_EXPORT` is causing a build error for deprecated `CVSReporter` with clang:

```cpp
class BENCHMARK_EXPORT BENCHMARK_DEPRECATED_MSG(       <-- HERE!
    "The CSV Reporter will be removed in a future release") CSVReporter
    : public BenchmarkReporter {

```
```js
In file included from ..\..\src\benchmark.cc:15:
../../include\benchmark/benchmark.h(1724): error : declaration of anonymous class must be a definition
../../include\benchmark/benchmark.h(1724): warning : declaration does not declare anything [-Wmissing-declarations]

```

Now, the details.
I want to build benchmark.dll, luckily, there is `BENCHMARK_EXPORT` annotation, but for .dll it's either incomplete or went to a wrong direction.

In essence `BENCHMARK_EXPORT` should expand to `__declspec(dllexport)` and for that I have to replace original export.h with my own:
```cpp
#if defined _WIN32 \
 || defined __CYGWIN__ \
 || defined __MSYS__				// or whatever else is emulated on Windows
#	if defined BENCHMARK_EXPORTS
#		define BENCHMARK_EXPORT		__declspec(dllexport)
#	elif defined BENCHMARK_SHARED
#		define BENCHMARK_EXPORT		__declspec(dllimport)
#	else
#		define BENCHMARK_EXPORT
#	endif
#	define BENCHMARK_NO_EXPORT
#else						// __linux__ or __APPLE__ or __ANDROID__
#	if defined BENCHMARK_EXPORTS
#		define BENCHMARK_EXPORT		__attribute__ ((visibility ("default")))
#		define BENCHMARK_NO_EXPORT	__attribute__ ((visibility ("hidden")))
#	else
#		define BENCHMARK_EXPORT
#		define BENCHMARK_NO_EXPORT
#	endif
#endif

```
Doing that produces an error since it seems that for clang `__attribute()`s and `__declspec()`s should be arranged properly: `__attribute()`s first, `__declspec()`s second and in the code we have EXPORT first and DEPRECATE second so it expands into reverse order.

To fix, `BENCHMARK_EXPORT` should be placed **after** `BENCHMARK_DEPRECATED_MSG()`:

```cpp
class BENCHMARK_DEPRECATED_MSG(
    "The CSV Reporter will be removed in a future release") BENCHMARK_EXPORT CSVReporter
    : public BenchmarkReporter {

```

**System**
Which OS, compiler, and compiler version are you using:
  - OS: Windows 10 21H2
  - Compiler and version: MSVC2017+, clang-cl (LLVM, 12.0)

**To reproduce**
1. Install LLVM toolchain for MSVC2017.
2. Run benchmark's cmake to generate export.h
3. Change export.h export part as above.
4. Observe the error.

**Additional context**
Alternatively, replacing `__attribute__((deprecated(msg)))` with `__declspec(deprecate(msg))` in `BENCHMARK_DEPRECATED_MSG()` (as it is done in **export.h** for `BENCHMARK_DEPRECATED`) would eliminate mixing `__attribute()`s and `__declspec()`s but would produce C4996 warning with cl.exe which Microsoft decided to treat as error, to ratify a compiler switch /w34996 could be used to force it back to be a level 3 warning. /Wd4996 compiler switch could be used to disable it instead of bulky `BENCHMARK_DISABLE/RESTORE_DEPRECATED_WARNING` or `#pragma`).
OR maybe retire `CVSReporter` already :).

**Comment by dominichamon** [comment]
i'd love to retire cvs reporter but i keep getting pushback ... v2.0 :)

if this is just a matter of swapping the order of macros in the csv reporter declaration, let's do that?

**Comment by sergiud** [comment]
I do not understand why are you replacing `export.h` by your own? What does the originally generated `export.h` contain? Probably not a `__declspec` but an `__attribute__`?

**Comment by yzazik** [comment]
@sergiud 
I have dozens of standalone entities (.dll to simplify) which are benchmark tested both individually and as a part of the whole, adding static linking to each one becomes a size issue since all of this should run on some dev aarch64 platform too.

Besides, many of them are explicitly (hard) linked into some other entities where it might get into already-defined-symbol conflicts, so I have my own .dll wrapper over benchmark.lib and it's becoming an issue maintaining it, so one of the solutions is to build benchmark.dll and delegate the integration/maintenance to a number of people who maintain those dozens of libraries.

export.h is missing __declspec(dllexport) declaration for exporting things in Windows build. Even if I would appeal to dev team here to support .dll/.so/.dyLib build, I do not really have high (and immediate) hopes of defending my needs while I'm under release delivery pressure. True, replacing export.h is only need for Windows, but Windows is where the most of the initial development takes place.

Other issue I have is growing API (so to speak, even if it's only by one file, ie export.h), the problem is that automation scripts expect certain things in certain place and each time there is a change it's all becoming a manual work.
Specifically, I only need a project to build benchmark library, all the rest in .sln is irrelevant, and since there is no separate dependent project that builds export.h I have to invent it myself, then code the script to cherry pick things from one place and construct in the other.

Why can't I use existent cmake flow? Well, I can but I still have to modify it, the build/sanity/test crons are running on virtual machine farms with read-only storage and redirectling input/output folders is what needs to be done and it is not really portable in stateless isolated environments (for example copy month old pre-built setup to another place and debug an issue, can't pull from github, can't recreate month old folder hierarchy from a different platform, etc)

In so many words, the reason is both necessity and convenience, adapting something to our existent build flow rather then adopting foreign is well established policy that I'm bound to.

This is what I get in generated export.h:

```cpp
#ifndef BENCHMARK_EXPORT_H
#define BENCHMARK_EXPORT_H

#ifdef BENCHMARK_STATIC_DEFINE
#  define BENCHMARK_EXPORT
#  define BENCHMARK_NO_EXPORT
#else
#  ifndef BENCHMARK_EXPORT
#    ifdef benchmark_EXPORTS
        /* We are building this library */
#      define BENCHMARK_EXPORT 
#    else
        /* We are using this library */
#      define BENCHMARK_EXPORT 
#    endif
#  endif

#  ifndef BENCHMARK_NO_EXPORT
#    define BENCHMARK_NO_EXPORT 
#  endif
#endif

#ifndef BENCHMARK_DEPRECATED
#  define BENCHMARK_DEPRECATED __declspec(deprecated)
#endif

#ifndef BENCHMARK_DEPRECATED_EXPORT
#  define BENCHMARK_DEPRECATED_EXPORT BENCHMARK_EXPORT BENCHMARK_DEPRECATED
#endif

#ifndef BENCHMARK_DEPRECATED_NO_EXPORT
#  define BENCHMARK_DEPRECATED_NO_EXPORT BENCHMARK_NO_EXPORT BENCHMARK_DEPRECATED
#endif

#if 0 /* DEFINE_NO_DEPRECATED */
#  ifndef BENCHMARK_NO_DEPRECATED
#    define BENCHMARK_NO_DEPRECATED
#  endif
#endif

#endif /* BENCHMARK_EXPORT_H */

```
In other words, not the things I need, but if `dllexport` will be added one day, then you would have the export-deprecated order issue in CVSReporter.

@dominichamon Would work for me as of now.

**Comment by sergiud** [comment]
OK, after finally getting through the whole text, I see many issues that you are causing yourself.

> export.h is missing __declspec(dllexport) declaration for exporting things in Windows build. 

That's only true if you have generated a static library variant of benchmark using CMake. Hence the export annotation is empty which is expected. Additionally, you should not be using the static library project files to build a shared library. Also note that there's no `BENCHMARK_EXPORTS` but there is `benchmark_EXPORTS` etc.

If you are duplicating the efforts for whatever reasons, you must to ensure that you correctly generate the actual project files and apply the same logic used by `GenerateExportHeader` which [will be updated in the future](https://gitlab.kitware.com/cmake/cmake/-/issues/22908).

**Comment by yzazik** [comment]
@sergiud 
> I see many issues that you are causing yourself

All of that was before export.h was introduced. After that, the substitution was simpler to deal with when building several platforms/targets from the same source location (otherwise things start to overwrite each other).

> That's only true if you have generated a static library variant of benchmark using CMake.

I did try to cmake with -DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=TRUE -DBUILD_SHARED_LIBS=TRUE, but then I get the order issue with clang.

**Comment by sergiud** [comment]
OK, that's fine. You can export all symbols. But then do you actually need a `__declspec` annotation? At least CMake detects this as not required if it's empty.

Could you also please attach your full command line you use to generate the MSVC clang project files using CMake including the CMake output?

**Comment by sergiud** [comment]
```console
-DCMAKE_CXX_COMPILER_ID=Clang
```
You should not be specifying the compiler ID. This is why CMake project generation fails since the subproject find MSVC 
```
-- The C compiler identification is MSVC 19.29.30139.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
```
but applies incorrect compiler identification
```console
  target_compile_features no known features for CXX compiler

  "Clang"
```
Try this instead:
```batch
set CC=clang
set CXX=clang++
cmake -DYOUR_FLAGS=...
```

**Comment by yzazik** [comment]
Still detects MSVC:

>set CC=clang
>set CXX=clang++
>echo %CC%
clang
>echo %CXX%
clang++
>cmake -E chdir "build" cmake -DBENCHMARK_DOWNLOAD_DEPENDENCIES=on -DCMAKE_BUILD_TYPE=Release -DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=TRUE -DBUILD_SHARED_LIBS=TRUE ../
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19044.
-- The CXX compiler identification is **MSVC 19.29.30139.0**
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done

**Comment by sergiud** [comment]
Forgot to mention that the above won't work with Visual Studio generator. You can use Ninja or nmake, or explicitly set the toolset to `llvm` or whatever the clang-cl identifier in Visual Studio is:
```console
cmake -T llvm -A x64 ...
```

**Comment by yzazik** [comment]
@dominichamon @sergiud 
This is going off topic, so what about swapping `BENCHMARK_EXPORT` and `BENCHMARK_DEPRECATED_MSG()` in `CVSReport`?

**Comment by sergiud** [comment]
The order of attributes should not matter. Since the problem is caused by your changes, you could simply replace your `__declspec` definition by `__attribute__` to avoid the issue.

**Comment by yzazik** [comment]
There is no `__atribute__((dllexport))`, there is only `__declspec(dllexport)`.

**Comment by sergiud** [comment]
I suspect Clang does provide `__attribute__ ((visibility(("default")))`.

**Comment by yzazik** [comment]
Well, with `__attribute__ ((visibility(("default")))` it builds but nothing gets exported.

**Comment by sergiud** [comment]
You did configure benchmark using `-DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=TRUE`, no? Then, additional export is not necessary.

@dominichamon Independently from the lengthy discussion here, I think it is more consistent to use `__declspec` style annotations on Windows even if `__attribute__` is available. Specifically, I would suggest updating this line
https://github.com/google/benchmark/blob/c5636440403c9790ee17d217ed511ce7467048e4/include/benchmark/benchmark.h#L245
and excluding `_WIN32` in the condition.

**Comment by yzazik** [comment]
@sergiud with EXPORT_ALL_SYMBOLS it works, but .dll is twice bigger (15 times more symbols) but I could probably live with that on Windows.
I still need to figure out how to force cmake to produce __attribute__ (or nothing) instead of __declspec for BENCHMARK_EXPORT if I want to keep original export.h. This even less convenient as it has an extra dependency on generated .def file.
Well, since proposed simpler solution is being disregarded, I guess, I'll find my way to deal with this. Thanks for all the help.

**Comment by sergiud** [comment]
The presumed workaround is not a viable solution because the order of attributes should not matter in general case. A different user could face a similar issue for which the order reversal could be suggested again coming full circle. A solution needs to generalize well.

You can try changing the above line to
```cpp
 #if defined(__GNUC__) || defined(__clang__)  && !defined(_WIN32)
```
and report back if it helps.

**Comment by yzazik** [comment]
@sergiud Agree, it shouldn't matter, but it does. Generic approach would be to put that into cmake's autocong flow and if that order does not build I guess you would decide what to do then.

`&& !defined(_WIN32)` does help, that basically makes `BENCHMARK_DEPRECATED_MSG()` to expand into nothing so __declspec and __attribute__ won't appear together.
Can't say it doesn't look like a different king of workaround, but works for me, I just thought that that deprecate message is "engraved in stone" so to speak.
